### PR TITLE
Merge entries with equivalent keys/masks

### PR DESCRIPTION
**Pull request by mundya**
This fixes a bug where entries with equivalent masks and keys were not merged. Routing table entries are now not constructed until the last possible minute, just after default entries are removed.

### CODE REVIEW

File: rig/place_and_route/utils.py [hunk 1]
```diff
@@ -73,8 +74,11 @@ def build_routing_tables(routes, net_keys, omit_default_routes=True):
     -------
     {(x, y): [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
     """
-    # {(x, y): [RoutingTableEntry, ...]
-    routing_tables = defaultdict(list)
+    # Pairs of inbound and outbound routes.
+    _InOutPair = namedtuple("_InOutPair", "ins, outs")
+
+    # (x, y) -> (key, mask) -> {_InOutPair, ...}
```
Comment (mossblaser) [inline rig/place_and_route/utils.py]: I don't understand this syntax(?)

File: rig/place_and_route/utils.py [hunk 2]
```diff
@@ -102,8 +106,29 @@ def build_routing_tables(routes, net_keys, omit_default_routes=True):
...
+                    set([direction]), set(out_directions)
+                )
+
+    # Construct the routing tables from the route sets
+    routing_tables = defaultdict(list)
+    for (x, y), routes in iteritems(route_sets):
+        for (key, mask), route in iteritems(routes):
+            # Remove default routes where possible
+            if omit_default_routes and (len(route.ins) == 1 and
+                                        route.outs == route.ins):
```
Comment (mossblaser) [inline rig/place_and_route/utils.py]: Shouldn't this check that the output route is *opposite* the input route, i.e. not the same(?)

File: rig/place_and_route/utils.py (hunk 1 above)
Comment (mundya) [inline rig/place_and_route/utils.py]: Sorry, "{(x, y): {(key, mask): [_InOutPair, ...], ...}, ...}"... Haskell leeches into my thinking sometimes...

File: rig/place_and_route/utils.py (hunk 2 above)
Comment (mundya) [inline rig/place_and_route/utils.py]: `ins` comes from `direction` (line 91 in this commit) which appears to be the direction out of the previous node.  (I copied you here).

Comment (mossblaser) [inline rig/place_and_route/utils.py]: Ah, yes! Jolly good! (It would have worried me more if the tests hadn't caught that if it was a bug ;)).

**Comment by mossblaser** [comment]
LGTM :) Thanks for that @mundya!

**Pull request merged**